packages/hash/shared/src/types.ts

```diff
@@ -0,0 +1,333 @@
+  },
+  icon: {
+    title: "Icon",
+    description: "An emoji icon.",
+  },
+  tokens: {
+    title: "Tokens",
+    description: undefined,
+  },
+} as const;
+
+type WorkspacePropertyTypeKey = keyof typeof workspacePropertyTypes;
+
+export type WorkspacePropertyTypeTitle =
+  typeof workspacePropertyTypes[WorkspacePropertyTypeKey]["title"];
+
+/**
+ * The workspace link type titles.
+ */
+const workspaceLinkTypes = {
+  hasMembership: {
+    title: "Has Membership",
+    description: "Having a membership.",
+  },
+  ofOrg: {
+    title: "Of Org",
+    description: "Belonging to an organization",
+  },
+  blockData: {
+    title: "Block Data",
+    description: "The entity representing the data in a block.",
+  },
+  contains: {
+    title: "Contains",
+    description: "Something containing something.",
+  },
+  parent: {
+    title: "Parent",
+    description: "The parent of something.",
+  },
+} as const;
+
+type WorkspaceLinkTypeKey = keyof typeof workspaceLinkTypes;
+
+export type WorkspaceLinkTypeTitle =
+  typeof workspaceLinkTypes[WorkspaceLinkTypeKey];
+
+/**
+ * The primitive data types ("Text", "Number", etc.)
+ */
+const primitiveDataTypes = {
+  text: {
+    title: "Text",
+    description: "An ordered sequence of characters",
+  },
+  number: {
+    title: "Number",
+    description: "An arithmetical value (in the Real number system)",
+  },
+  boolean: {
+    title: "Boolean",
+    description: "A True or False value",
+  },
+  emptyList: {
+    title: "Empty List",
+    description: "An Empty List",
+  },
+  object: {
+    title: "Object",
+    description: "A plain JSON object with no pre-defined structure",
+  },
+  null: {
+    title: "Null",
+    description: "A placeholder value representing 'nothing'",
+  },
+} as const;
+
+export type PrimitiveDataTypeKey = keyof typeof primitiveDataTypes;
+
+export type PrimitiveDataTypeTitle =
+  typeof primitiveDataTypes[PrimitiveDataTypeKey]["title"];
+
+type TypeDefinition = {
+  title: string;
+  description?: string;
+};
+
+type EntityTypeDefinition = TypeDefinition & { entityTypeId: VersionedUri };
+
+type PropertyTypeDefinition = TypeDefinition & { propertyTypeId: VersionedUri };
+
+type DataTypeDefinition = TypeDefinition & { dataTypeId: VersionedUri };
+
+type LinkTypeDefinition = TypeDefinition & {
+  linkTypeId: VersionedUri;
+  description: string;
+};
+
+type TypeDefinitions = {
+  entityType: Record<WorkspaceEntityTypeKey, EntityTypeDefinition>;
+  propertyType: Record<WorkspacePropertyTypeKey, PropertyTypeDefinition>;
+  linkType: Record<WorkspaceLinkTypeKey, LinkTypeDefinition>;
+  dataType: Record<PrimitiveDataTypeKey, DataTypeDefinition>;
+};
+
+/**
+ * The workspace and block protocol types that are statically available at run-time.
+ *
+ * @todo - instead of just providing the id, title and description of types, find a way
+ * to make the entire schemas statically available at run-time.
+ */
+export const types: TypeDefinitions = {
```

Comment:
@thehabbos007 @Alfred-Mountfield do you agree with this `@todo`? In which case I suppose I should capture it in a (de-prioritised) asana task and link it here

Reply:
I think we do want the entire schemas statically available, yes. I wonder if using the `@blockprotocol/type-system-web` types would be easier than using this custom-made `TypeDefinition` type? I believe most of the metadata should be accessible from the package without needing WASM (thus usable in shared)

Reply:
>  I wonder if using the @blockprotocol/type-system-web types would be easier than using this custom-made TypeDefinition type?

@thehabbos007 can you be more specific? `TypeDefinition` does not re-define the schema type definitions, it's just a collection of `title`/`description`/identifiers, so I'm not sure what you're suggesting.

Something like this?
```TypeScript
type TypeDefinitions = {
  entityType: Record<WorkspaceEntityTypeKey, Omit<PersistedEntityType, "properties" | "links" | // ...eveything else that isn't an id, title or description>;
  ...
};
```

Reply:
> I think we could get static typing if we somehow infer the const structure of our workspace types against these definitions, but not sure how trivial that is

Happy to take suggestions, but I can't think of a simple way to set this up, and I don't think we'd get any typing improvements in the end result (as `title` and `description` just have type `string`, and `id` already has `VersionedUri` as its type in what I've setup here). Unless there's some other benefit I'm missing from deriving these types from that package, I'm not sure it's worth worrying about until we actually make the schemas statically available where the typing becomes more complex.

Reply:
I'm not _sure_ I'd capture it in an asana task at this point or keep the todo. I think there's a wider problem we'll need to solve, how to define types more ergonomically, make them discoverable, etc.

I don't think that just applies to the workspace types and I wouldn't want to see us picking up this task just because there is a task, and then to waste work in it without solving the bigger problem.

I'm not sure though.

Reply:
Since I'm also not sure what exactly the solution will look like, and whether it will be addressed in the `shared` package at all, I've removed the todo as you suggested in https://github.com/hashintel/hash/pull/1193/commits/ce0e32243dad13f27408ad6fdfa1fea94009da91